cleanrl/ppo_continuous_action_envpool_jax.py

```diff
@@ -0,0 +1,377 @@
+        frac = 1.0 - (count // (args.num_minibatches * args.update_epochs)) / args.num_updates
+        return args.learning_rate * frac
+    agent_optimizer = optax.chain(
+        optax.clip_by_global_norm(args.max_grad_norm),
+        optax.inject_hyperparams(optax.adam)(learning_rate=linear_schedule, eps=1e-5),
+    )
+    agent_params = AgentParams(
+        actor_params,
+        critic_params,
+    )
+    agent_optimizer_state = agent_optimizer.init(agent_params)
+
+    # ALGO Logic: Storage setup
+    obs = jnp.zeros((args.num_steps, args.num_envs) + envs.single_observation_space.shape)
+    actions = jnp.zeros((args.num_steps, args.num_envs) + envs.single_action_space.shape)
+    logprobs = jnp.zeros((args.num_steps, args.num_envs))
+    rewards = np.zeros((args.num_steps, args.num_envs))
+    dones = jnp.zeros((args.num_steps, args.num_envs))
+    values = jnp.zeros((args.num_steps, args.num_envs))
+    advantages = jnp.zeros((args.num_steps, args.num_envs))
+    avg_returns = deque(maxlen=20)
+
+    @jax.jit
+    def get_action_and_value(x, d, obs, dones, actions, logprobs, values, step, agent_params, key):
+        obs = obs.at[step].set(x)  # inside jit() `x = x.at[idx].set(y)` is in-place.
+        dones = dones.at[step].set(d)
+        action_mean, action_logstd = actor.apply(agent_params.actor_params, x)
+        # action_logstd = (jnp.ones_like(action_mean) * action_logstd)
+        action_std = jnp.exp(action_logstd)
+        key, subkey = jax.random.split(key)
+        action = action_mean + action_std * jax.random.normal(subkey, shape=action_mean.shape)
+        logprob = -0.5 * ((action - action_mean) / action_std) ** 2 - 0.5 * jnp.log(2.0 * jnp.pi) - action_logstd
+        entropy = action_logstd + 0.5 * jnp.log(2.0 * jnp.pi * jnp.e)
+        value = critic.apply(agent_params.critic_params, x)
+        actions = actions.at[step].set(action)
+        logprobs = logprobs.at[step].set(logprob.sum(1))
+        values = values.at[step].set(value.squeeze())
+        return obs, dones, actions, logprobs, values, action, logprob, entropy, value, key
+
+    @jax.jit
+    def get_action_and_value2(x, action, agent_params):
+        action_mean, action_logstd = actor.apply(agent_params.actor_params, x)
+        action_std = jnp.exp(action_logstd)
+        logprob = -0.5 * ((action - action_mean) / action_std) ** 2 - 0.5 * jnp.log(2.0 * jnp.pi) - action_logstd
+        entropy = action_logstd + 0.5 * jnp.log(2.0 * jnp.pi * jnp.e)
+        value = critic.apply(agent_params.critic_params, x).squeeze()
+        return logprob.sum(1), entropy, value
+
+    @jax.jit
+    def compute_gae(next_obs, next_done, rewards, dones, values, advantages, agent_params):
+        advantages = advantages.at[:].set(0.0)  # reset advantages
+        next_value = critic.apply(agent_params.critic_params, next_obs).squeeze()
+        lastgaelam = 0
+        for t in reversed(range(args.num_steps)):
```

Comment:
Was looking through your codes to get some idea about how other people were writing PPO/SAC (and how far people jited things) and think this might be an issue during the first compile step. The for loop will basically be unrolled and when I tried this the compile time was very long.

Ended using `jax.lax.scan`

```
    not_dones = ~dones
    
    value_diffs = gamma * values[1:] * not_dones - values[:-1]
    deltas = rewards + value_diffs

    def body_fun(gae, t):
        gae = deltas[t] + gamma * gae_lambda * not_dones[t] * gae
        return gae, gae
    indices = jnp.arange(N)[::-1]
    gae, advantages = jax.lax.scan(body_fun, 0.0, indices,)
    advantages = advantages[::-1]
 ```
Also avoids using the .at and .set functions (of which im still not sure of what the performance is).

Reply:
you can use `reverse=True` in the scan so you don't have to flip it.